src/core/server/http/http_server.ts

```diff
@@ -163,43 +163,54 @@ export class HttpServer {
         const validate = isSafeMethod(route.method) ? undefined : { payload: true };
         const { authRequired, tags, body = {}, timeout } = route.options;
         const { accepts: allow, maxBytes, output, parse } = body;
-        // Hapi does not allow timeouts on payloads to be specified for 'head' or 'get' requests
-        const payloadTimeout = isSafeMethod(route.method) || timeout == null ? undefined : timeout;
 
         const kibanaRouteState: KibanaRouteState = {
           xsrfRequired: route.options.xsrfRequired ?? !isSafeMethod(route.method),
         };
 
+        // Works around https://github.com/hapijs/hapi/issues/4122 until v20
+        const fakeSocketTimeout = timeout?.payload ? timeout.payload + 1 : undefined;
+
         this.server.route({
           handler: route.handler,
           method: route.method,
           path: route.path,
           options: {
             auth: this.getAuthOption(authRequired),
             app: kibanaRouteState,
+            ext: {
+              onPreAuth: {
+                method: (request, reply) => {
+                  if (timeout?.idleSocket) {
+                    request.raw.req.socket.setTimeout(timeout?.idleSocket);
+                  } else if (fakeSocketTimeout) {
+                    request.raw.req.socket.setTimeout(this.config!.socketTimeout || 0);
```

Comment:
It would be helpful for us later if there was a comment explaining how we set the initial socket timeout to be the fake one, then override it here to ensure that it reflects the configuration. At first glance I thought this code was wrong until I really read the details of the Hapi bug.

It may also be useful to add a comment that `0` means no timeout.

I would also suggest using `??` instead so that we can avoid a bug if `this.config!.socketTimeout` is set to 0 and we change the default to a non-zero. In that case, using `||` would ignore the configured timeout and always use the default. Maybe overly defensive but why not?

```suggestion
                    request.raw.req.socket.setTimeout(this.config!.socketTimeout ?? 0);
```

Reply:
> It would be helpful for us later if there was a comment explaining how we set the initial socket timeout to be the fake one, then override it here to ensure that it reflects the configuration. At first glance I thought this code was wrong until I really read the details of the Hapi bug.

> It may also be useful to add a comment that 0 means no timeout.

:+1:

> I would also suggest using ?? instead so that we can avoid a bug if this.config!.socketTimeout is set to 0 and we change the default to a non-zero. In that case, using || would ignore the configured timeout and always use the default. Maybe overly defensive but why not?

Good catch! I forgot that TS added this null-ish coalescing... It's truly the future!!!